core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala

```diff
@@ -16,48 +16,110 @@
+  def toWriteOperation(operation: WriteOperation): v1_2.WriteOperation = v1_2.WriteOperation(
+    outputSource = operation.outputSource,
+    append = operation.append,
+    id = operation.id,
+    name = Some(operation.name),
+    childIds = operation.childIds,
+    params = toUntypedMap(operation.params).asNonEmptyOption,
+    extra = toUntypedMap(operation.extra).asNonEmptyOption
   )
 
-  override def toDTO(event: ExecutionEvent): Option[v1_2.ExecutionEvent] = Some(automap(event).to[v1_2.ExecutionEvent])
+  def toReadOperation(operation: ReadOperation): v1_2.ReadOperation = v1_2.ReadOperation(
+    inputSources = operation.inputSources,
+    id = operation.id,
+    name = Some(operation.name),
+    output = operation.output.asNonEmptyOption,
+    params = toUntypedMap(operation.params).asNonEmptyOption,
+    extra = toUntypedMap(operation.extra).asNonEmptyOption
+  )
+
+  def toDataOperation(operation: DataOperation): v1_2.DataOperation = v1_2.DataOperation(
+    id = operation.id,
+    name = Some(operation.name),
+    childIds = operation.childIds.asNonEmptyOption,
+    output = operation.output.asNonEmptyOption,
+    params = toUntypedMap(operation.params).asNonEmptyOption,
+    extra = toUntypedMap(operation.extra).asNonEmptyOption
+  )
 
-  implicit def toAttrOrExprRef(o: Seq[AttrOrExprRef]): Option[Seq[v1_2.AttrOrExprRef]] = o.map {
+  def toAttribute(attribute: Attribute): v1_2.Attribute = v1_2.Attribute(
+    id = attribute.id,
+    dataType = attribute.dataType,
+    childRefs = attribute.childRefs.map(toAttrOrExprRef).asNonEmptyOption,
+    extra = toUntypedMap(attribute.extra).asNonEmptyOption,
+    name = attribute.name
+  )
+
+  def toAttrOrExprRef(o: AttrOrExprRef): v1_2.AttrOrExprRef = o match {
     case AttrRef(attrId) => v1_2.AttrOrExprRef(Some(attrId), None)
     case ExprRef(exprId) => v1_2.AttrOrExprRef(None, Some(exprId))
   }
+
+  val untypedMapConverter = new UntypedMapConverter(toAttrOrExprRef)
+  def toUntypedMap(map: Map[String, Any]): Map[String,Any] = untypedMapConverter.toUntypedMap(map)
+
+  def toExpressions(expressions: Expressions): v1_2.Expressions = v1_2.Expressions(
+    functions = expressions.functions.map(toFunctionalExpression).asNonEmptyOption,
+    constants = expressions.constants.map(toConstant).asNonEmptyOption
+  )
+
+  def toFunctionalExpression(fe: FunctionalExpression): v1_2.FunctionalExpression = v1_2.FunctionalExpression(
+    id = fe.id,
+    dataType = fe.dataType,
+    childRefs = fe.childRefs.map(toAttrOrExprRef).asNonEmptyOption,
+    extra = toUntypedMap(fe.extra).asNonEmptyOption,
+    name = fe.name,
+    params = toUntypedMap(fe.params).asNonEmptyOption
+  )
+
+  def toConstant(l: Literal): v1_2.Literal = v1_2.Literal(
+    id = l.id,
+    dataType = l.dataType,
+    extra = toUntypedMap(l.extra).asNonEmptyOption,
+    value = l.value
+  )
+
+  def toNameAndVersion(nav: NameAndVersion): v1_2.NameAndVersion = v1_2.NameAndVersion(
+    name = nav.name, version = nav.version
+  )
```

Comment:
formatting: either write the entire call statement in one line, or place one arg per line.